src/Bedrock.Framework/Protocols/MessagePipeReader.cs

```diff
@@ -0,0 +1,179 @@
+﻿using System;
+using System.Buffers;
+using System.IO;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Bedrock.Framework.Protocols
+{
+    public class MessagePipeReader : PipeReader
+    {
+        private readonly IMessageReader<ReadOnlySequence<byte>> _messageReader;
+        private readonly PipeReader _reader;
+
+        private SequencePosition _examined;
+        private SequencePosition _consumed;
+        private ReadOnlySequence<byte> _message;
+        private bool _isThisCompleted;
+        private bool _isCanceled;
+        private bool _isCompleted;
+        private readonly ConsumableArrayBufferWriter<byte> _backlog = new ConsumableArrayBufferWriter<byte>();
+        private bool _allExamined;
+
+        public MessagePipeReader(PipeReader reader, IMessageReader<ReadOnlySequence<byte>> messageReader)
+        {
+            _reader = reader ?? throw new ArgumentNullException(nameof(reader));
+            _messageReader = messageReader ?? throw new ArgumentNullException(nameof(messageReader));
+        }
+
+        public override void AdvanceTo(SequencePosition consumed)
+        {
+            AdvanceTo(consumed, consumed);
+        }
+
+        public override void AdvanceTo(SequencePosition consumed, SequencePosition examined)
+        {
+            if (_isThisCompleted)
+            {
+                ThrowReadAfterCompleted();
+            }
+
+            if (_message.Slice(examined).Length == 0)
```

Comment:
```suggestion
            if (_message.Slice(examined).IsEmpty)
```
Though more a matter of taste.